daemon/rm.go

```diff
@@ -74,3 +77,78 @@ func (daemon *Daemon) RemovePodContainer(p *Pod) {
 	}
 	daemon.db.DeleteP2C(p.Id)
 }
+
+func (daemon *Daemon) DeleteContainer(containerId string) error {
+	pod, idx, ok := daemon.PodList.GetByContainerIdOrName(containerId)
+	if !ok {
+		return fmt.Errorf("can not find container %s", containerId)
+	}
+
+	if !pod.TransitionLock("rm") {
+		glog.Errorf("Pod %s is under other operation", pod.Id)
+		return fmt.Errorf("Pod %s is under other operation", pod.Id)
+	}
+	defer pod.TransitionUnlock("rm")
+
+	if pod.status.Status == types.S_POD_RUNNING && pod.status.Containers[idx].Status == types.S_POD_RUNNING {
+		err := daemon.StopContainerWithinLock(pod, containerId)
+		if err != nil {
+			return fmt.Errorf("failed to stop container %s", containerId)
+		}
+	}
+
+	pod.Lock()
+	defer pod.Unlock()
+
+	pod.Status().DeleteContainer(containerId)
+
+	daemon.PodList.Put(pod)
+	if err := daemon.WritePodAndContainers(pod.Id); err != nil {
+		glog.Errorf("Found an error while saving the Containers info: %v", err)
+		return err
+	}
+
+	r, err := daemon.ContainerInspect(containerId, false, version.Version("1.21"))
+	if err != nil {
+		return err
+	}
+
+	if err := daemon.Daemon.ContainerRm(containerId, &dockertypes.ContainerRmConfig{}); err != nil {
+		return fmt.Errorf("failed to rm container %s", containerId)
+	}
+
+	rsp, ok := r.(*dockertypes.ContainerJSON)
+	if !ok {
+		err = fmt.Errorf("fail to unpack container json response for %s of %s", containerId, pod.Id)
+		return err
+	}
+	name := strings.TrimLeft(rsp.Name, "/")
+	for i, c := range pod.spec.Containers {
+		if name == c.Name {
+			pod.spec.Containers = append(pod.spec.Containers[:i], pod.spec.Containers[i+1:]...)
+			break
+		}
+	}
+	podSpec, err := json.Marshal(pod.spec)
+	if err != nil {
+		glog.Errorf("Marshal podspec %v failed: %v", pod.spec, err)
+		return err
+	}
+	if err = daemon.db.UpdatePod(pod.Id, podSpec); err != nil {
+		glog.Errorf("Found an error while saving the POD file: %v", err)
+		return err
+	}
+
+	jsons, err := pod.TryLoadContainers(daemon)
+	if err != nil {
+		return err
+	}
+	if err = pod.ParseContainerJsons(daemon, jsons); err != nil {
+		glog.Errorf("Found an error while parsing the Containers json: %v", err)
+		return err
+	}
+
+	daemon.PodList.Put(pod)
```

Comment:
podlist.Put updates the containers array in podlist while daemon.WritePodAndContainers uses podstatus.Containers to updates the database, the podstatus.Containers was already updated.